src/trace/layers/mod.rs

```diff
@@ -106,6 +109,73 @@ pub trait Cursor<Storage> {
     fn reposition(&mut self, storage: &Storage, lower: usize, upper: usize);
 }
 
+/// A general-purpose container resembling `Vec<T>`.
+pub trait BatchContainer: Default {
+    /// The type of contained item.
+    type Item;
+    /// Inserts an owned item.
+    fn push(&mut self, item: Self::Item);
+    /// Inserts a borrowed item.
+    fn copy(&mut self, item: &Self::Item);
+    /// Extends from a slice of items.
+    fn copy_slice(&mut self, slice: &[Self::Item]);
+    /// Creates a new container with sufficient capacity.
+    fn with_capacity(size: usize) -> Self;
+    /// Reserves additional capacity.
+    fn reserve(&mut self, additional: usize);
+    /// Creates a new container with sufficient capacity.
+    fn merge_capacity(cont1: &Self, cont2: &Self) -> Self;
+}
+
+impl<T: Clone> BatchContainer for Vec<T> {
+    type Item = T;
+    fn push(&mut self, item: T) {
+        self.push(item);
+    }
+    fn copy(&mut self, item: &T) {
+        self.push(item.clone());
+    }
+    fn copy_slice(&mut self, slice: &[T]) {
+        self.extend_from_slice(slice);
+    }
+    fn with_capacity(size: usize) -> Self {
+        Vec::with_capacity(size)
+    }
+    fn reserve(&mut self, additional: usize) {
+        self.reserve(additional);
+    }
+    fn merge_capacity(cont1: &Self, cont2: &Self) -> Self {
+        Vec::with_capacity(cont1.len() + cont2.len())
+    }
+}
+
+impl<T: Columnation> BatchContainer for TimelyStack<T> {
+    type Item = T;
+    fn push(&mut self, item: T) {
+        self.copy(&item);
+    }
+    fn copy(&mut self, item: &T) {
+        self.copy(item);
+    }
+    fn copy_slice(&mut self, slice: &[T]) {
+        self.reserve_items(slice.iter());
+        for item in slice.iter() {
+            self.copy(item);
+        }
+    }
+    fn with_capacity(size: usize) -> Self {
+        Self::with_capacity(size)
+    }
+    fn reserve(&mut self, _additional: usize) {
+    }
```

Comment:
We use this in various `push_merge` implementations for trie layers. I don't think it should block merging this PR, but probably worth pondering at least briefly. For example:
```rust
        let (trie1, mut lower1, upper1) = other1;
        let (trie2, mut lower2, upper2) = other2;

        self.vals.reserve((upper1 - lower1) + (upper2 - lower2));
```